experimental/type-checker/src/KindChecker.hs

```diff
@@ -0,0 +1,339 @@
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
+
+module KindChecker (getType) where
+
+import Control.Monad.Trans.Reader (ReaderT, runReaderT)
+import Control.Monad.Trans.State (StateT, evalStateT)
+import Control.Monad.Trans.Writer (WriterT (runWriterT))
+import Data.Bifunctor (Bifunctor (second))
+
+import Control.Monad.Except
+import Control.Monad.Reader.Class
+import Control.Monad.State.Class
+import Control.Monad.Writer.Class
+
+import Prettyprinter
+
+type Atom = String
+infixr 8 :->:
+
+data Kind
+  = Kind
+  | Kind :->: Kind
+  | KVar String
+  deriving (Eq, Show)
+
```

Comment:
Can't we just use `Natural` (everywhere)? 

While I appreciate that we want to keep this open in case we want to eventually support kinds like `Nat` or `Symbol`, the changes needed to support that will be massive/breaking no matter what we do - e.g. the codegen can safely assume for now that every TyVar arguments to a type represents something of kind `*`, if we add `Nat` or `Symbol` kinds it's not even clear what we'd *do* in e.g. Rust or JS. 

But anyway, both of your versions of `Kind` allow things like: 

``` haskell 
k1,k2,k3,k4 :: Kind 
k1 = KindRef "a" 
k2 = KindRef "b"
k3 = KindRef "c" 
k4 = KindRef "d" 
ty = KindRef "Type"

pattern (:->) :: Kind -> Kind -> Kind 
pattern (:->) a b = KindArrow a b 

bad :: Kind 
bad = ((((k1 :-> k2) :-> k3) :-> ty) :-> k4) :-> ty 
```

If y'all just don't like `Natural`s here, I guess we could do something like 

``` haskell 
data Star 

data Arrow a b  

data Kind :: forall k. k -> Type where 
  TyKind :: Kind Star 
  (:->) :: forall k. Kind Star -> Kind k -> Kind (Arrow Star k)
infixr 5 :->

ty :: Kind Star
ty = TyKind

-- compiles
good :: Kind (Arrow Star (Arrow Star Star))
good = ty :-> ty :-> ty

{-  • Couldn't match type ‘Arrow Star Star’ with ‘Star’
      Expected: Kind Star
        Actual: Kind (Arrow Star Star)
    • In the first argument of ‘(:->)’, namely ‘(ty :-> ty)’
      In the expression: (ty :-> ty) :-> ty
      In an equation for ‘bad’: bad = (ty :-> ty) :-> ty
-}
bad = (ty :-> ty) :-> ty
```

But I suggest we just go with `Natural`.